In [1]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_pickle('화장품리뷰all.plk')

df.head(n=2)

# 만들어놓은 데이터 병합

In [ ]:
화장품정보 = pd.read_pickle('화장품정보all.plk')
화장품정보.head(n=2)

In [ ]:
화장품최종본 = pd.read_pickle('화장품최종본all.plk')
화장품최종본['리뷰수'] = 화장품최종본['리뷰수'].str.replace(",","").str.extract('([0-9]+)')[0].astype(int)

In [ ]:
화장품최종본.head(n=2)

In [ ]:
cols0 = ['브랜드','이름','가격','종류']
cols1 = ['브랜드','이름','평균리뷰점수','리뷰수','링크']

In [ ]:
df1 = pd.merge(left = 화장품최종본[cols1], right = 화장품정보[cols0], how='inner', on = ['브랜드','이름'])

In [ ]:
df1.head()

In [ ]:
완성본 = pd.merge(left = df, right = df1, how = 'left', left_on = '상품이름' ,right_on='이름')

In [ ]:
완성본.head(1)

In [ ]:
완성본2 = 완성본.drop(columns = ['링크','이름'], axis=1)

In [ ]:
완성본2.columns

In [ ]:
# 1. 원하는 순서대로 컬럼 리스트 정의
new_columns_order = [
    '브랜드', '상품이름', '가격', '종류', '평균리뷰점수', 
    '리뷰수', '작성자', '별점', '피부타입', '날짜', 
    '옵션', '리뷰내용', '상품링크'
]

# 2. 데이터프레임에 적용 (df는 현재 가지고 계신 데이터프레임 변수명)
oy_data = 완성본2[new_columns_order]



In [ ]:
oy_data.head()

가격 타입바꾸고
날짜 타입바꾸고

In [ ]:
oy_data['가격'] = oy_data['가격'].astype(dtype = 'int')
oy_data['날짜']  = oy_data['날짜'].astype(dtype = 'datetime64[ns]')

자료수집날짜 12/28일 기준 약 23937개

In [74]:
df = pd.read_pickle('oy_data.plk')

In [3]:
df['브랜드'].value_counts()

브랜드
오브제     11086
비레디      5784
듀이셀      1782
다슈       1376
원오브뎀     1162
그라펜      1151
라끌랑       971
스웨거       263
두잉왓       217
정샘물       115
엠도씨        30
Name: count, dtype: int64

# 리뷰 중복 제거
한 상품 내에서 동일인이 리뷰를 비슷한 내용으로 2개 이상 쓰거나, 기획상품과 원래 상품의 리뷰가 동일한 경우가 발생 이를 제거하자

그러면 몇 개정도가 그럴까?

중복이 생기는 이유
1. 반복 구매 및 재구매
2. 옵션별 리뷰 작성
3. CJ포인트 적립목적

In [32]:
중복전체 = df[df.duplicated(['작성자','리뷰내용_정제'], keep=False)]
중복전체.groupby(['작성자','리뷰내용_정제'])['상품이름'].nunique().value_counts()


상품이름
2    6265
3      32
1       9
4       1
5       1
Name: count, dtype: int64

In [34]:
중복전체[
    중복전체['상품이름'].isin([
        '듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)',
        '듀이셀 트루빔 파운데이션 12g 기획'
    ])
][['작성자','리뷰내용_정제','상품이름']]


작성자                                            리뷰내용_정제  \
11518        고주부  남성용 파운데이션을 살 때 가장 중요한 점은 편리함이라고 생각합니다 올인원 제품이라...   
11519       코코천사  대학생아들이 자연스러운 파운데이션을 사달라고 해 스틱형 파운데이션이면서 브러쉬까지 ...   
11520     새콤달콤캔디  색상이 자연스러워서 얼굴에 발랐을 때 둥둥 뜨지 않아요 스틱 피운데이션을 얼굴에 쓱...   
11521        누를황  남자도 관리하는 시대로 남자 메이크업 제품도 좋은게 많이 나오더라구요 아닌 분도 많...   
11522       눈을깜자  브러쉬 덮개가 있어서 덮개를 내리면 안에 브러쉬가 있는 형태라서 케이스에 넣다가 브...   
...          ...                                                ...   
12446    tnw****                        추천받아서 구매햇는데 좋은거같아요 ㅎㅎ 괜찮습니다   
12447   ggok****  항상 잘쓰고 있습니다 또 재구매했습니다 트러블 없이 피부에도 잘 맞아서 계속 쓰게 ...   
12448  leehe****       지인에게 추천받아서 사용해보는데 생각보다 제품도 좋고 저한테 잘 맞는 것 같아요   
12449    tnw****                    촉촉하고 괜찮은 것 같아요 저렴하게 구매해서 좋아욘ㅎㅎㅎ   
12450   zoon****  광고보고 구매했습니다 생각보다 자연스럽게 나오는데 피부톤 23호 이상으로 밝으면 약...   

                                  상품이름  
11518  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  
11519  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  
11520  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  
11521  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  
11522  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  
...                                ...  
12446  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  
12447  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  
12448  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  
12449  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  
12450  듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)  

[848 rows x 3 columns]

In [75]:

# 리뷰에서 특수문자, 이모지 제거
df['리뷰내용_정제'] = (
    df['리뷰내용']
    .str.replace(r'[^\w\s]', '', regex=True)
    .str.replace(r'\s+', ' ', regex=True)
    .str.strip()
)

중복데이터 = df.loc[df.duplicated(['작성자','리뷰내용_정제']),:]

In [ ]:
중복데이터2 = df.loc[df.duplicated(['리뷰내용_정제']),:]

In [ ]:
len(set(df.loc[df['상품이름']=='듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)']['리뷰내용']) - set(df.loc[df['상품이름']=='듀이셀 트루빔 파운데이션 12g']['리뷰내용']))

0

In [47]:
df.loc[:,'상품이름'].value_counts()

상품이름
오브제 내추럴 커버 파운데이션 기획세트(+클렌징폼 50ml)                   6452
[덱스 PICK] 오브제 내추럴 커버 파운데이션 13g (SPF50+, PA++++)     3677
비레디 블루 쿠션 4세대 15g(SPF 34 PA++)                      3251
비레디 블루 쿠션 4세대 리필 15g                                1677
[하루종일 착붙] 원오브뎀 드 뗑 쿠션 12g(1호, 2호)                   1162
[덱스 PICK]오브제 퍼펙트 옴므 쿠션 15g (SPF50+PA+++)             957
듀이셀 트루빔 파운데이션 12g 기획(+토너패드 20매)                      935
비레디 블루 파운데이션 35ml                                    856
듀이셀 트루빔 파운데이션 12g                                    847
다슈 맨즈 멀티 커버 스틱 파운데이션 12g 뉴트럴베이지 기획(+미니 톤 컨트롤 로션)     643
다슈 맨즈 멀티 커버 스틱 파운데이션 ( 단품/기획 )                       643
[맨즈파데1등] 그라펜 팁 스케치 멀티 커버 파운데이션 20ml                  494
라끌랑 슈퍼쉴드 옴므쿠션 2종 택1                                  454
[착붙쿠션] 그라펜 핸섬 커버쿠션 2종 택1                             430
라끌랑 내츄럴 커버 옴므쿠션 3종 택1                                414
[올인원 완벽커버] 스웨거 톤 커버 파운데이션 스틱 [1호/2호]                 263
[피부착붙컬러] 그라펜 매치업 파운데이션 40ml 단품/기획                    227
두잉왓 소프트 블러링 필터 쿠션 15g 단품/기

In [21]:
df.duplicated(subset=['리뷰내용_정제'], keep='first').sum()

np.int64(6379)

작성자와 리뷰내용이 중복인 데이터의 행수와
리뷰내용만 중복인 행수가 21개 차이.

이것의 원인은 작성자는 다른데 리뷰내용이 똑같기 때문에 발생

즉, 다른사람의 리뷰를 복사해서 자신의 리뷰로 사용한 것으로 판단.

이게 매크로일 수도 있고 아닐 수도있는데

이들을 찾아보자

In [76]:
# 1. 내용만 중복인 인덱스 (더 넓은 범위)
idx_content_only = df.duplicated(subset=['리뷰내용_정제'], keep='first')

# 2. 작성자+내용 모두 중복인 인덱스 (더 좁은 범위)
idx_author_content = df.duplicated(subset=['작성자', '리뷰내용_정제'], keep='first')

# 3. 두 조건의 차이가 발생하는 데이터 추출
# (내용은 중복인데, 작성자가 달라서 '작성자+내용' 기준에서는 중복으로 안 잡힌 데이터)
diff_df = df[idx_content_only & ~idx_author_content]

print(f"내용만 중복인 수: {idx_content_only.sum()}")
print(f"작성자+내용 중복인 수: {idx_author_content.sum()}")
print(f"차이나는 행 수: {len(diff_df)}")

# 차이 데이터 확인
diff_df[['작성자', '리뷰내용_정제', '상품이름']].head()

내용만 중복인 수: 6379
작성자+내용 중복인 수: 6358
차이나는 행 수: 21


작성자                                            리뷰내용_정제  \
4279   kimgeo****                 트러블 없이 잘 맞아서 민감성 피부에도 무난하게 쓸 수 있어요   
8076     lees****  올리브영은 참 다양한 상품을 팔아서 좋아요 먹거리 부터 화장품부터 전자제품까지 정말...   
9236      tlf****  다음엔 할인 더 해줬으면 좋겠습니다 ㅠㅠ 많이많이 구매할게요 너무너무 대만족입니다 ...   
10473  glorio****  너무너무 만족합니다 지인추천으로 구매하게 되었고 잘사용하고 있습니다 대만족 저도 지...   
10612  alsdnd****  한 달 동안 사용하며 흐름이 자연스럽게 이어지는 점을 가장 크게 체감했습니다 불편한...   

                                           상품이름  
4279          오브제 내추럴 커버 파운데이션 기획세트(+클렌징폼 50ml)  
8076             비레디 블루 쿠션 4세대 15g(SPF 34 PA++)  
9236             비레디 블루 쿠션 4세대 15g(SPF 34 PA++)  
10473                         비레디 블루 파운데이션 35ml  
10612  [덱스 PICK]오브제 퍼펙트 옴므 쿠션 15g (SPF50+PA+++)

In [ ]:
적은데이터 = df.loc[df['리뷰내용_정제'].str.len()<20,:]

그래서 위의 데이터를 어떻게 처리하냐??

21개를 그냥 중복으로 보고 아예 제거하고 사용할까?

생각해본결과 어차피 매크로든 복붙이든 자기가 쓴 리뷰가 아니므로 삭제하기로 결정

즉, 리뷰 내용이 같으면 그냥 삭제

거기다가 '좋아요' 같은 단순 리뷰는 존재하지 않기때문에 제거해도 될 것 같음



다슈맨즈 멀티커버 스틱 파운데이션 기획이랑 일반이랑 리뷰 완전 똑같음.
일반 상품 제거
다슈 맨즈 멀티 커버 스틱 파운데이션 12g 뉴트럴베이지 기획(+미니 톤 컨트롤 로션)

단 가격은 원가로 측정?

듀이셀도 같은게 많네요
듀이셀은 기획은 남겨두고 나머지 제거

1. 안전하게 다슈맨즈 멀티커버스틱파운데이션, 듀이셀 일반상품 제거
2. 그리고 drop_duplicates로 중복제거
3. 자카드 유사도로 최종적으로 약간만 다르게한 데이터 제거

In [62]:
df = df.loc[~df['상품이름'].isin(['다슈 맨즈 멀티 커버 스틱 파운데이션 ( 단품/기획 )',
        '듀이셀 트루빔 파운데이션 12g'])]

In [64]:
df2 = df.drop(columns='리뷰내용')

In [65]:
df_정제 = df2.drop_duplicates('리뷰내용_정제')

23937개 데이터에서 17558개가 되었음 딱 6379개 제거

duplicated로 정제해도 완전히 똑같지 않으면 제거가 되지 않아서 자카드 유사도 이용해서 제거

보통 1달 후기 리뷰같은 경우는 기존 리뷰에서 단어 조금만 바꿔서 올리기 때문에 이와같은 과정을 거쳐야함

In [77]:
import pandas as pd
from itertools import combinations

# 자카드 유사도 함수 (공백 무시)
def get_jaccard_sim(str1, str2):
    if not str1 or not str2: return 0
    a = set(str1.replace(" ", ""))
    b = set(str2.replace(" ", ""))
    return len(a & b) / len(a | b) if len(a | b) != 0 else 0

# 1. 날짜를 기준으로 정렬 (최신 리뷰를 남기고 싶으면 ascending=False)
df = df.sort_values(by=['작성자', '날짜'])

indices_to_drop = set()

# 2. 작성자별로 그룹화하여 비교
for author, group in df.groupby('작성자'):
    idx_list = group.index.tolist()
    if len(idx_list) < 2: continue  # 리뷰가 하나면 통과
    
    # 그룹 내에서 두 개씩 짝지어 비교
    for i, j in combinations(idx_list, 2):
        sim = get_jaccard_sim(df.loc[i, '리뷰내용_정제'], df.loc[j, '리뷰내용_정제'])
        
        # 유사도가 0.8(80%) 이상이면 중복으로 판단
        if sim >= 0.8:
            # i가 이전 날짜이므로 j(나중 날짜)를 삭제 (원하는 기준에 따라 선택)
            indices_to_drop.add(j)

# 3. 최종 데이터 생성
df_final = df.drop(index=list(indices_to_drop))

print(f"기존 데이터: {len(df)}개")
print(f"유사 중복 제거 후: {len(df_final)}개")

기존 데이터: 23937개
유사 중복 제거 후: 17463개


In [72]:
df_final.to_pickle('중복데이터제거.plk')

최종 중복제거데이터 17444개

In [73]:
pd.read_pickle('중복데이터제거.plk')

브랜드                                              상품이름     가격  \
8652    비레디                    비레디 블루 쿠션 4세대 15g(SPF 34 PA++)  32900   
22456    다슈  다슈 맨즈 멀티 커버 스틱 파운데이션 12g 뉴트럴베이지 기획(+미니 톤 컨트롤 로션)  29000   
22638   두잉왓                       두잉왓 소프트 블러링 필터 쿠션 15g 단품/기획  33900   
2728    오브제                 오브제 내추럴 커버 파운데이션 기획세트(+클렌징폼 50ml)  30100   
2714    오브제                 오브제 내추럴 커버 파운데이션 기획세트(+클렌징폼 50ml)  30100   
...     ...                                               ...    ...   
21822   그라펜                 [피부착붙컬러] 그라펜 매치업 파운데이션 40ml 단품/기획  24000   
19443  원오브뎀                 [하루종일 착붙] 원오브뎀 드 뗑 쿠션 12g(1호, 2호)  26500   
20262  원오브뎀                 [하루종일 착붙] 원오브뎀 드 뗑 쿠션 12g(1호, 2호)  26500   
19408  원오브뎀                 [하루종일 착붙] 원오브뎀 드 뗑 쿠션 12g(1호, 2호)  26500   
20260  원오브뎀                 [하루종일 착붙] 원오브뎀 드 뗑 쿠션 12g(1호, 2호)  26500   

            종류  평균리뷰점수   리뷰수         작성자  별점           피부타입         날짜  \
8652        쿠션     4.8  3323     000****   5           None 2025-10-19   
22456  스틱파운데이션     4.8   671    0000****   5           None 2024-10-15   
22638       쿠션     4.7   219  0009999999   5  지성,겨울쿨톤,잡티,미백 2024-03-12   
2728   스틱파운데이션     4.8  6892  0009999999   5       건성,쿨톤,미백 2024-09-06   
2714   스틱파운데이션     4.8  6892  0009999999   5       건성,쿨톤,미백 2024-10-09   
...        ...     ...   ...         ...  ..            ...        ...   
21822    파운데이션     4.8   230          힙플   5           None 2024-08-31   
19443       쿠션     5.0  1169         힝코쥬   5           None 2025-03-30   
20262       쿠션     5.0  1169         힝코쥬   5           None 2025-05-01   
19408       쿠션     5.0  1169         힝코쥬   5           None 2025-08-16   
20260       쿠션     5.0  1169         힝코쥬   5           None 2025-09-19   

                           옵션  \
8652              [옵션] [본품]1호   
22456                    None   
22638          [옵션] 1호 라이트베이지   
2728   [옵션] 2호 기획(+클렌징폼 50ml)   
2714   [옵션] 2호 기획(+클렌징폼 50ml)   
...                       ...   
21822                    None   
19443                [옵션] 001   
20262                [옵션] 001   
19408                [옵션] 001   
20260                [옵션] 001   

                                                    상품링크  \
8652   https://www.oliveyoung.co.kr/store/goods/getGo...   
22456  https://www.oliveyoung.co.kr/store/goods/getGo...   
22638  https://www.oliveyoung.co.kr/store/goods/getGo...   
2728   https://www.oliveyoung.co.kr/store/goods/getGo...   
2714   https://www.oliveyoung.co.kr/store/goods/getGo...   
...                                                  ...   
21822  https://www.oliveyoung.co.kr/store/goods/getGo...   
19443  https://www.oliveyoung.co.kr/store/goods/getGo...   
20262  https://www.oliveyoung.co.kr/store/goods/getGo...   
19408  https://www.oliveyoung.co.kr/store/goods/getGo...   
20260  https://www.oliveyoung.co.kr/store/goods/getGo...   

                                                 리뷰내용_정제  
8652   제가 원래 얼굴이 밝은 편인데 이번 여름에 신나게 놀아서 그런건지 몰라도 원래 쓰던...  
22456  제가 여자인데 개똥손이라 이거 쓰는데 피부 밝으신 분들은 밝은 선크림 한 번 파운데...  
22638  좋아하는 유튜버분 브랜드라 더 기대되는 쿠션이에요 ㅎ 내돈내산할까 고민하던 찰나 체...  
2728   선물용으로 구입했어요 우선 스틱 타입이라 손에 묻을 일도 없고 정말 편해보였습니다 ...  
2714   무난하게 커버 잘 되고 브러쉬도 부드러워요 손에 안 묻고 편하게 슥슥 바르면 되니 ...  
...                                                  ...  
21822  진짜 매번 쓰는 제품인데요 제 피부톤이랑 딱 맞아서 너무 자연스러워서 늘 만족합니다...  
19443  여자친구한테 잘보이려고 구매한 쿠션 일반적인 로션제품들은 손에 너무 뭍어나서 바를때...  
20262                       남자친구가 완전 만족해하면서 사용하는중이에요 굿굿굿  
19408  남자친구가 요즘 화장하는거에 맛들렸는지 이것저것 사다가 이제는 하다하다 쿠션을 사달...  
20260         사주니까 야무지게 잘사용하고있어서 뿌듯 할때마다 피부가 뽀용해보여서 대리만족  

[17444 rows x 13 columns]

In [3]:
import pandas as pd